src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -1495,6 +1497,161 @@ static bool extendedTypeIsOrDerivedFrom(NamedTypeSymbol hiddenContainer, NamedTy
+
+                removeLessSpecificNonMethodCurrentSymbols(additionalSymbol, currentSymbols, out bool isAdditionalSymbolHidden,
+                    currentCount, basesBeingResolved, ref useSiteInfo);
+
+                if (!isAdditionalSymbolHidden)
+                {
+                    currentSymbols.Add(additionalSymbol);
+                }
+            }
+
+            return;
+
+            // Determine whether we want to add the additional symbol
+            // and potentially remove some "hidden" non-method current symbols in the process
+            void removeLessSpecificNonMethodCurrentSymbols(Symbol additionalSymbol, ArrayBuilder<Symbol> currentSymbols, out bool isAdditionalSymbolHidden,
+                int currentCount, ConsList<TypeSymbol> basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                isAdditionalSymbolHidden = false;
+                for (int j = 0; j < currentCount; j++)
+                {
+                    var currentSymbol = currentSymbols[j];
+                    if (currentSymbol is null)
+                    {
+                        continue;
+                    }
+
+                    if (currentSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } currentContainerUnderlying
+                        || additionalSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } additionalContainerUnderlying)
+                    {
+                        // We wouldn't have gathered these members if we didn't have an extended type for them.
+                        throw ExceptionUtilities.Unreachable();
+                    }
+
+                    if (additionalContainerUnderlying.Equals(currentContainerUnderlying, TypeCompareKind.ConsiderEverything))
+                    {
+                        // neither underlying type is more specific
+                    }
+                    else if (IsDerivedType(baseType: additionalContainerUnderlying, derivedType: currentContainerUnderlying,
+                        basesBeingResolved, this.Compilation, ref useSiteInfo))
+                    {
+                        // current symbol is more specific, it "hides" a less specific non-method/non-indexer additional symbol
+                        if (!IsMethodOrIndexer(additionalSymbol))
+                        {
+                            isAdditionalSymbolHidden = true;
+                        }
+                    }
+                    else if (IsDerivedType(baseType: currentContainerUnderlying, derivedType: additionalContainerUnderlying,
+                        basesBeingResolved, this.Compilation, ref useSiteInfo))
+                    {
+                        // additional symbol is more specific, it "hides" a less specific non-method/non-indexer current symbol
+                        if (!IsMethodOrIndexer(currentSymbol))
+                        {
+                            currentSymbols[j] = null;
```

Comment:
>currentSymbols[j] = null; [](http://example.com/codeflow?start=28&length=25)

I do not think we should be modifying an ArrayBuilder owned by LookupResult this way.